File: 2017/04/04.py
def preprocessing(data):
    return (passphrase.split(' ') for passphrase in data.splitlines())


def solver(passphrases):
    valid       = 0
    safer_valid = 0
    
    for passphrase in passphrases:
        if len(passphrase) == len(set(passphrase)): 
            valid += 1
            if len(set(''.join(sorted(p)) for p in passphrase)) == len(passphrase):
                safer_valid += 1
                
    yield valid
    yield safer_valid

File: 2017/04/04.rs
aoc::main!();

fn preprocessing(input: &str) -> Vec<Vec<&str>> {
    input
    .lines()
    .map(|line| 
        line.split(' ').collect_vec())
    .collect_vec()
}

fn part_1(passphrases: Vec<Vec<&str>>) -> usize {
    passphrases
    .iter()
    .filter(|passphrase| 
        passphrase
        .iter()
        .all_unique())
    .count()
}

fn part_2(passphrases: Vec<Vec<&str>>) -> usize {
    passphrases
    .iter()
    .filter(|passphrase| 
        passphrase
        .iter()
        .map(|p| p.chars().sorted().collect_vec())
        .all_unique())
    .count()
}